### ISSUE TITLE
unable to build librrpage

### ISSUE THREAD

**Issue by liangjs** [rr-debugger/rr#2854]
```
[ 24%] Linking C shared library lib/rr/librrpage_32.so
/usr/bin/cmake -E cmake_link_script CMakeFiles/rrpage_32.dir/link.txt --verbose=1
/usr/bin/cc -fPIC  -D__USE_LARGEFILE64 -pthread -msse2 -D__MMX__ -D__SSE__ -D__SSE2__ -Wstrict-prototypes -std=gnu11  -pthread -g -m32 -Wl,-T -Wl,/home/liangjs/third-party/rr/src/preload/rr_page.ld -nostartfiles -nostdlib -fuse-ld=bfd -shared -Wl,-soname,librrpage_32.so -o lib/rr/librrpage_32.so CMakeFiles/rrpage_32.dir/32/preload/rr_page.S.o 
../src/preload/tweak_librrpage.py /home/liangjs/third-party/rr/build/lib/rr/librrpage_32.so
Traceback (most recent call last):
  File "../src/preload/tweak_librrpage.py", line 117, in <module>
    assert (alloc_offset - new_section_offset) <= reserved_size
AssertionError
```

This is the failed `librrpage_32.so`.
[librrpage_32.zip](https://github.com/rr-debugger/rr/files/6457375/librrpage_32.zip)

**Comment by khuey** [comment]
@Keno

**Comment by Keno** [comment]
Try this patch:
```
diff --git a/src/preload/tweak_librrpage.py b/src/preload/tweak_librrpage.py
index 41139fda..820dd964 100755
--- a/src/preload/tweak_librrpage.py
+++ b/src/preload/tweak_librrpage.py
@@ -97,7 +97,9 @@ with open(sys.argv[1], 'rb+') as f:
        write_uptr(is64, f, new_table_offset)
 
        alloc_offset = new_table_offset + size
-       for n in range(12, 15):
+
+        # Move .symtab, .strtab, .shstrtab into the pre-allocated space (.sh_placeholder)
+       for n in range(e_shnum - 3, e_shnum):
                seek_nth_section_sh_offset(f, new_table_offset, e_shentsize, n, sh_offset_offset)
                sh_offs = read_uptr(is64, f)
                sh_size = read_uptr(is64, f)
```

**Comment by liangjs** [comment]
It works. Thanks!

**Comment by khuey** [comment]
Let's get this pushed before we close the issue :)

**Comment by khuey** [comment]
This got landed as b774becea808bdc4fe558390429846a66e3a5324
